service/kvm_config_resources_v2.go

```diff
@@ -0,0 +1,192 @@
+	return kvmConfigResourcesV2Config{
+		CertsSearcher:      nil,
+		K8sClient:          nil,
+		Logger:             nil,
+		RandomkeysSearcher: nil,
+
+		Name: "",
+	}
+}
+
+func newKVMConfigResourcesV2(config kvmConfigResourcesV2Config) ([]framework.Resource, error) {
+	var err error
+
+	if config.CertsSearcher == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.CertsSearcher must not be empty")
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+	if config.RandomkeysSearcher == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.RandomkeysSearcher must not be empty")
+	}
+
+	if config.Name == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Name must not be empty")
+	}
+
+	var ccServiceV2 *cloudconfigv2.CloudConfig
+	{
+		c := cloudconfigv2.DefaultConfig()
+
+		c.Logger = config.Logger
+
+		ccServiceV2, err = cloudconfigv2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var namespaceResource framework.Resource
+	{
+		c := namespacev2.DefaultConfig()
+
+		c.K8sClient = config.K8sClient
+		c.Logger = config.Logger
+
+		namespaceResource, err = namespacev2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var configMapResourceV2 framework.Resource
+	{
+		c := configmapv2.DefaultConfig()
+
+		c.CertSearcher = config.CertsSearcher
+		c.CloudConfig = ccServiceV2
+		c.K8sClient = config.K8sClient
+		c.Logger = config.Logger
+
+		configMapResourceV2, err = configmapv2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var deploymentResourceV2 framework.Resource
+	{
+		c := deploymentv2.DefaultConfig()
+
+		c.K8sClient = config.K8sClient
+		c.Logger = config.Logger
+
+		deploymentResourceV2, err = deploymentv2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var ingressResource framework.Resource
+	{
+		c := ingressv2.DefaultConfig()
+
+		c.K8sClient = config.K8sClient
+		c.Logger = config.Logger
+
+		ingressResource, err = ingressv2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var pvcResource framework.Resource
+	{
+		c := pvcv2.DefaultConfig()
+
+		c.K8sClient = config.K8sClient
+		c.Logger = config.Logger
+
+		pvcResource, err = pvcv2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var serviceResource framework.Resource
+	{
+		c := servicev2.DefaultConfig()
+
+		c.K8sClient = config.K8sClient
+		c.Logger = config.Logger
+
+		serviceResource, err = servicev2.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	resources := []framework.Resource{
+		namespaceResource,
+
+		configMapResourceV2,
+		deploymentResourceV2,
+		ingressResource,
+		pvcResource,
+		serviceResource,
+	}
+
+	// Wrap resources with retry and metrics.
+	{
```

Comment:
What's the lifetime of V2 vs. V3 (I mean how long these separated codes do exist before cleanup)? Could common parts (between v2 and v3) such as this wrapping be extracted to single function that's used here then? I could imagine things like backoff parameters would have changes at some point in the future and maintenance would be easier and less error prone if the code is not duplicated. WDYT?

Reply:
Yes good shout. I was also thinking about it. In general we should aim to remove previous versions quickly. This is only a matter of a customer running an old version. It was stated that we will encourage customers to upgrade quickly.

Therefore I don't think duplication is an issue. We should work only on a latest version unless there is a bug. I think it will be more of a problem if we want to change something in a single version w/o touching other. So I think in this case duplication is actually good.